Clear db connection after database error

heyrutvik opened the pull request:
Closes #3342

heyrutvik commented:
The build is failed because of [this](https://github.com/diesel-rs/diesel/actions/runs/3126043831/jobs/5071080266). I'm not seeing any clippy errors in local but will make changes as per the build log.

seeekr commented on diesel_tests/tests/connection.rs:
```diff
@@ -2,22 +2,21 @@ use crate::schema::*;
 use diesel::connection::BoxableConnection;
 use diesel::*;
 
-table! {
```
Moving around & changing code that's not necessary as part of the new or updated test code should be avoided, i.e. usually the minimal set of changes to resolve an issue is preferable.
If this is indeed updating some test code to follow newer and better patterns that are already in use elsewhere, that's of course acceptable and probably desirable. I don't know the specifics of this case, but I wanted to leave this general note so you could act on it based on your best judgment.

heyrutvik replied:
@seeekr the PR has been updated to keep only ticket related changes. Thanks for your feedback.

heyrutvik commented:
@weiznich There is a [table definition](https://github.com/diesel-rs/diesel/blob/master/diesel_tests/tests/connection.rs#L5-L11) in the connection.rs file which could be moved to the relevant test method. I don't see a reason to keep it outside. The test I added [encloses table definition within](https://github.com/diesel-rs/diesel/pull/3344/files#diff-fe8a7db272b54bc3a035d79dca6f87144fbeb215cd460a481afcabee27b98557R127-R133) it which is inspired from other test cases, [example](https://github.com/diesel-rs/diesel/blob/master/diesel/src/pg/connection/mod.rs#L736).

@seeekr has made a suggestion that we can move that table definition to the relevant test method (and other similar pattern in the codebase, if any) using following changeset to keep the changes minimal,
```rust
fn managing_updated_at_for_table() {

    table! {
        #[sql_name = "auto_time"]
            auto_time_table {
            id -> Integer,
            n -> Integer,
            updated_at -> Timestamp,
        }
    }

    use auto_time_table::dsl::{auto_time_table as auto_time, n, updated_at};
    ...
```

Initially, I made changes to enclose that table definition, different from this, but it was cluttering the actual changes. I have reverted it to push only changes relevant to the ticket. Let me know your thoughts about it, whether we want to include such changes in current PR or in a separate one.

weiznich commented on diesel/src/pg/connection/result.rs:
```diff
@@ -48,6 +48,12 @@ impl PgResult {
+                // this will indicate that the previous command is complete and
+                // the same connection is ready to process next command.
+                // https://www.postgresql.org/docs/current/libpq-async.html
+                while conn.get_next_result()?.is_some() {}
```
We want to return the orginal error here, so we want to avoid using `?` at all. Something like

```suggestion
                while conn.get_next_result().map(Option::is_some).unwrap_or(true) {}
```

(although this will recursively call the same function (`PgResult::new`) again and potentially run this loop there again. I believe that should be fine as this should stop at some layer wen `get_next_result()` returns `None`)

heyrutvik replied:
@weiznich Is the default value `true` intended? Wouldn't it keep the loop running even after some error? Am I missing something?

weiznich replied:
Yes `true` is the correct default value as far as I can tell. The libpq documentation states that `PQgetResult` (which is called internally by this function) must be called till it returns `null`. It does not talk about any errors there, so I assume that means we literally need to call it again till it returns null.

That by itself seems reasonable as [`get_next_result()`](https://github.com/diesel-rs/diesel/blob/master/diesel/src/pg/connection/raw.rs#L119) calls `PQgetResult` (which cannot return an error) and then `PgResult::new()` (which can return an error). So the only way `get_next_result()` would ever return an error is by hitting basically the current code block. We would like to ignore that as the first error message is relevant.

weiznich reviewed (commented):
Thanks for filling this PR :+1:
This looks already good, but I feel that we should act a bit more careful here in terms of error propagation.

heyrutvik commented:
Added separate commit to move table definition to its relevant test method. I haven't looked into all the tests file because of the time crunch but updated the same file where I added the test for the ticket. Also, updated the polling code as per your suggestion. Thanks, @weiznich.

weiznich commented:
Thanks again for providing the example + working on the fix :+1: I will try to fix https://github.com/diesel-rs/diesel/issues/3330 as well and then put all the recent fixes into a 2.0.1 patch release. I cannot give an ETA on the exact release date.
